File: python/src/aps_kit/adapter.py
"""Runner adapter: persistent NDJSON worker for gherkin-mutator.

Each input line is a JSON job request; each output line is a JSON job response.
Diagnostics go to stderr only.
"""
from __future__ import annotations

import json
import os
import subprocess
import sys
import time
from pathlib import Path
from typing import Any


def _classify(returncode: int) -> str:
    if returncode == 0:
        return "test_success"
    if returncode == 1 or returncode == 2 or returncode == 3 or returncode == 4 or returncode == 5:
        # pytest exit codes: 1 (tests failed), 2 (interrupted), 3 (internal), 4 (usage), 5 (no tests).
        # Treat 1 as test_failure and the rest as infrastructure_error.
        return "test_failure" if returncode == 1 else "infrastructure_error"
    return "infrastructure_error"


def _parse_timeout(s: str | None) -> float | None:
    if not s:
        return None
    s = s.strip()
    if s.endswith("ms"):
        return float(s[:-2]) / 1000.0
    if s.endswith("s"):
        return float(s[:-1])
    if s.endswith("m"):
        return float(s[:-1]) * 60.0
    try:
        return float(s)
    except ValueError:
        return None


def _handle(job: dict[str, Any], test_command: list[str], cwd: Path | None) -> dict[str, Any]:
    env = os.environ.copy()
    env["APS_IR_PATH"] = job["feature_json"]
    env["APS_GENERATED_DIR"] = job.get("generated_dir", "")
    env["APS_WORK_DIR"] = job.get("work_dir", "")
    timeout = _parse_timeout(job.get("timeout"))
    started = time.monotonic_ns()
    try:
        proc = subprocess.run(
            test_command,
            env=env,
            cwd=cwd,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
    except subprocess.TimeoutExpired as exc:
        elapsed = time.monotonic_ns() - started
        return {
            "id": job["id"],
            "outcome": "infrastructure_error",
            "output": exc.stdout or "",
            "error": f"timeout after {timeout}s",
            "duration": elapsed,
        }
    elapsed = time.monotonic_ns() - started
    return {
        "id": job["id"],
        "outcome": _classify(proc.returncode),
        "output": proc.stdout,
        "error": proc.stderr,
        "duration": elapsed,
    }


def serve(test_command: list[str], cwd: Path | None = None) -> int:
    """Loop reading NDJSON jobs from stdin, writing NDJSON responses to stdout."""
    for raw in sys.stdin:
        raw = raw.strip()
        if not raw:
            continue
        try:
            job = json.loads(raw)
        except json.JSONDecodeError as exc:
            print(f"aps-adapter: bad job line: {exc}", file=sys.stderr)
            continue
        try:
            resp = _handle(job, test_command, cwd)
        except Exception as exc:  # noqa: BLE001
            resp = {
                "id": job.get("id", ""),
                "outcome": "infrastructure_error",
                "output": "",
                "error": f"adapter exception: {exc}",
                "duration": 0,
            }
        sys.stdout.write(json.dumps(resp) + "\n")
        sys.stdout.flush()
    return 0

File: typescript/src/adapter.ts
// Runner adapter: persistent NDJSON worker for gherkin-mutator.
import { spawnSync } from "node:child_process";
import { createInterface } from "node:readline";
import type { Readable, Writable } from "node:stream";

interface JobRequest {
  id: string;
  feature_json: string;
  generated_dir?: string;
  work_dir?: string;
  timeout?: string;
}

interface JobResponse {
  id: string;
  outcome: "test_success" | "test_failure" | "infrastructure_error";
  output: string;
  error: string;
  duration: number;
}

// Public so tests pin classification semantics without subprocess setup.
export function classifyExit(
  status: number | null,
): JobResponse["outcome"] {
  // Spec contract: exit 0 = test_success, exit 1 = test_failure, anything
  // else (including timeout) = infrastructure_error.
  if (status === 0) return "test_success";
  if (status === 1) return "test_failure";
  return "infrastructure_error";
}

// Public so tests pin duration parsing without subprocess setup.
export function parseTimeoutMs(s: string | undefined): number | undefined {
  if (!s) return undefined;
  const trimmed = s.trim();
  if (trimmed.endsWith("ms")) return Number(trimmed.slice(0, -2));
  if (trimmed.endsWith("s")) return Number(trimmed.slice(0, -1)) * 1000;
  if (trimmed.endsWith("m")) return Number(trimmed.slice(0, -1)) * 60_000;
  const n = Number(trimmed);
  return Number.isFinite(n) ? n * 1000 : undefined;
}

export interface ServeOptions {
  command: string;
  args: string[];
  cwd?: string;
}

/** Real-world entry point: uses process.stdin / stdout / stderr. */
export async function serve(opts: ServeOptions): Promise<void> {
  await serveIO(opts, process.stdin, process.stdout, process.stderr);
}

/** Testable entry point: takes injectable streams so unit tests can drive
 *  the NDJSON protocol in-process. */
export async function serveIO(
  opts: ServeOptions,
  input: Readable,
  output: Writable,
  diagnostics: Writable,
): Promise<void> {
  const rl = createInterface({ input });
  for await (const line of rl) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    let job: JobRequest;
    try {
      job = JSON.parse(trimmed);
    } catch (err) {
      diagnostics.write(
        `aps-adapter: bad job line: ${(err as Error).message}\n`,
      );
      continue;
    }
    const response = runOne(job, opts);
    output.write(JSON.stringify(response) + "\n");
  }
}

function runOne(job: JobRequest, opts: ServeOptions): JobResponse {
  const start = process.hrtime.bigint();
  const env = {
    ...process.env,
    APS_IR_PATH: job.feature_json,
    APS_GENERATED_DIR: job.generated_dir ?? "",
    APS_WORK_DIR: job.work_dir ?? "",
  };
  const proc = spawnSync(opts.command, opts.args, {
    env,
    cwd: opts.cwd,
    encoding: "utf-8",
    timeout: parseTimeoutMs(job.timeout),
  });
  const duration = Number(process.hrtime.bigint() - start);
  if (proc.error && proc.signal === "SIGTERM") {
    return {
      id: job.id,
      outcome: "infrastructure_error",
      output: proc.stdout ?? "",
      error: `timeout: ${proc.error.message}`,
      duration,
    };
  }
  if (proc.error) {
    return {
      id: job.id,
      outcome: "infrastructure_error",
      output: proc.stdout ?? "",
      error: proc.error.message,
      duration,
    };
  }
  return {
    id: job.id,
    outcome: classifyExit(proc.status),
    output: proc.stdout ?? "",
    error: proc.stderr ?? "",
    duration,
  };
}
